libsplinter/src/admin/service/open_proposals.rs

```diff
@@ -135,14 +135,32 @@ impl ProposalRegistry {
         }
     }
 
-    pub fn get_proposals(&self) -> Vec<messages::CircuitProposal> {
-        self.proposals
-            .iter()
-            .map(|(_, proposal)| proposal.clone())
-            .collect()
+    pub fn get_proposals(&self) -> Proposals {
+        Proposals {
+            inner: Box::new(self.proposals.clone().into_iter()),
+            size: self.proposals.len(),
+        }
     }
 
     pub fn has_proposal(&self, circuit_id: &str) -> bool {
         self.proposals.contains_key(circuit_id)
     }
 }
+
+/// An iterator over CircuitPropoals and the time that each occurred.
+pub struct Proposals {
+    inner: Box<dyn Iterator<Item = (String, messages::CircuitProposal)>>,
+    size: usize,
+}
+
+impl Iterator for Proposals {
+    type Item = (String, messages::CircuitProposal);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
```

Comment:
This should certainly stay, but we should just expose that `size` field from the Proposals struct via a function called `total`, that way it will be clear from it's usage. In fact, I might like to change `Events` to the same (in a future PR).